beps/0010-scaffolder-templating-in-parameters/README.md

```diff
@@ -0,0 +1,212 @@
+
+<!--
+List the specific goals of the BEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- This BEP will settle the implementation for the templating of fields into the JSON Schema in the `parameters` section in the scaffolder templates.
+- This BEP will settle how to pass through templating strings to underlying field extensions in a non-conflicting way.
+
+### Non-Goals
+
+<!--
+What is out of scope for this BEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to be able to decorate the template schema server side with a context and use that to drive the form rendering client side.
+
+We can extend the `/parameter-schema` endpoint to accept a `formData` context query parameter which will be a JSON object of the current `formData` state.
+
+```diff
+export interface ScaffolderApi {
+  getTemplateParameterSchema(
+    templateRef: string,
++    formData?: string,
+  ): Promise<TemplateParameterSchema>;
+}
+```
+
+```diff
+ router
+    .get(
+      '/v2/templates/:namespace/:kind/:name/parameter-schema',
+      async (req, res) => {
+        const credentials = await httpAuth.credentials(req);
+        const { token } = await auth.getPluginRequestToken({
+          onBehalfOf: credentials,
+          targetPluginId: 'catalog',
+        });
+        const template = await authorizeTemplate(
+          req.params,
+          token,
+          credentials,
+        );
+
+        const parameters = [template.spec.parameters ?? []].flat();
++        const secureTemplater = await SecureTemplater.loadRenderer({
++          templateFilters: {
++            ...createDefaultFilters({ integrations }),
++            ...additionalTemplateFilters,
++          },
++          templateGlobals: additionalTemplateGlobals,
++        });
++
++        const templatedParameters = parameters.map(parameter =>
++          renderTemplateString(
++            parameter,
++            {
++              parameters: req.query.formData,
++            },
++            secureTemplater,
++            logger,
++          ),
++        );
+```
+
+You can see a quick implementation of this in this [branch](https://github.com/backstage/backstage/compare/master...blam/templating-in-parameters)
+
+## Design Details
+
+There's a slight issue with the implementation of the `react-jsonschema-form`, which makes things like live updating on things like the `default` field slightly more difficult.
+Currently, on first render, the default value is populated and then stored in the `formData` object or the current state, and the default value is never re-evaluated again at a later stage.
+
+This means that if end users are wanting to set default values with `${{ parameters.myOtherProperty }}`, then they would need to ensure that they are on different steps in the form
+as the form would need to be re-rendered, and for performance reasons, we don't want to re-render the form on every `formData` update.
```

Comment:
are debounced rerenders possible?

Reply:
Hmm, I'm not sure it helps with the overall problem though to be honest, I think it's just going to be a limitation that the default value is populated straight away on first render, not on submit. But maybe there's another way around this that we can work out later.